Fix deadlock when disposing consumer combined with the consumerdispatcher

Rotvig opened the pull request:
- Reworked the ConsumerDispatcher to improve scenarios around dispose and OnDisconnect.
- Fixed the deadlocked that happened on a Fail-over combined with Dispose of consumer from the InernalConsumer. The schedueld dispose messages was removed from the queue OnDisconnect and the AutoResetEvent was never signaled in the InternalConsumer Dispose.
- Added Cancel to HandleModelShutdown in the BasicConsumer, so subscriber of consumer event will know if the consumer is shutdown by the model

zidad commented:
Interesting work! I don't fully understand the part with the priorities, isn't it possible to just use the priorities that are built-in RabbitMQ instead? I think this might be too confusing and too customized, perhaps we can add it as a separate assembly? I also have some specialized internal classes with more specialized logic which you can just replace in the DI container, I'd prefer to keep it out of the core product if you don't mind?

zidad commented on Source/EasyNetQ/Consumer/ConsumerDispatcher.cs:
```diff
@@ -31,7 +32,7 @@ public ConsumerDispatcher(ConnectionConfiguration configuration)
                         else
                         {
-                            await Task.Delay(50);
+                            manualResetEvent.WaitOne();
```
nice... I think the MRE should be disposed somewhere though

Rotvig replied:
True - not sure where it should be disposed. Becuase i think the queue should run until they are empty.

sungam3r replied:
In principle, nothing bad will happen if you do not dispose MRE.

sungam3r commented on Source/EasyNetQ/Consumer/ConsumerDispatcher.cs:
```diff
@@ -1,29 +1,38 @@
             var thread = new Thread(_ =>
             {
-                while (!disposed && queue.TryTake(out var action, -1))
+                while(!IsDone())
```
```suggestion
                while (!IsDone())
```

sungam3r commented on Source/EasyNetQ/Consumer/ConsumerDispatcher.cs:
```diff
@@ -1,29 +1,38 @@
+        private readonly ConcurrentQueue<Action> mediumPriority = new ConcurrentQueue<Action>();
+        private readonly ConcurrentQueue<Action> lowPriority = new ConcurrentQueue<Action>();
+        private readonly AutoResetEvent autoResetEvent = new AutoResetEvent(false);
         private bool disposed;
```
```suggestion
        private volatile bool disposed;
```

sungam3r commented on Source/EasyNetQ/Consumer/ConsumerDispatcher.cs:
```diff
@@ -34,25 +43,49 @@ public ConsumerDispatcher(ConnectionConfiguration configuration)
-            }
+                // throw away any queued actions. RabbitMQ will redeliver any in-flight
+                // messages that have not been acked when the connection is lost.
+                while (lowPriority.TryDequeue(out _))
```
Why do you clear only one queue from three existing?

sungam3r replied:
Also why do you perform `TryDequeue` not direct but through `QueueAction` call?

sungam3r commented on Source/EasyNetQ/Consumer/ConsumerDispatcher.cs:
```diff
@@ -34,25 +43,49 @@ public ConsumerDispatcher(ConnectionConfiguration configuration)
+
+        public void QueueAction(Action action, Priority priority = Priority.Low)
         {
             Preconditions.CheckNotNull(action, "action");
```
```suggestion
            Preconditions.CheckNotNull(action, "action");
            if (disposed)
                throw new ObjectDisposedException(nameof(ConsumerDispatcher));
```

This check will be equivalent for `queue.CompleteAdding();` which disallow `queue.Add(action);` after being called.

sungam3r commented:
> isn't it possible to just use the priorities that are built-in RabbitMQ instead? I think this might be too confusing and too customized, perhaps we can add it as a separate assembly? 

RabbitMQ priorities is completely different thing. I really like the idea with different queues. This is generally the right approach to streamline the handling of independent delegates. But errors are possible, since now it is possible to change the order of calling delegates compared to the order that was before the changes. I am talking about implicit dependencies that are already present in the code. Delegates can change the shared state. Now all delegates are in the same collection. As soon as someone decides to use priorities by passing a new parameter, the order of work with the shared state is changed.

This **may** or **may not** lead to errors. I understand that this is not the problem of the dispatcher itself, but of the code inside the delegates which should not make assumptions about the order of the calls.

Rotvig commented:
@sungam3r @zidad really good feeback ! i have removed the priority concept and added Durable and Transient. Could people confuse it with Rabbit related subjects ? or is it clear what the intention of the queues are ?

sungam3r commented on Source/EasyNetQ/Consumer/IConsumerDispatcher.cs:
```diff
@@ -4,7 +4,18 @@ namespace EasyNetQ.Consumer
 {
     public interface IConsumerDispatcher : IDisposable
     {
-        void QueueAction(Action action);
```
Breaking change. You could leave `QueueAction` name for transient case (+ of course documenting this) and add only new `QueueDurableAction` method.

sungam3r replied:
I personally like your option too.

Rotvig replied:
I am up for both. I prefer breaking the interface on this one - to make it very clear what you are doing. @zidad what is you oppinion on this ? :)

sungam3r replied:
I’ll just remind that this is still a breaking change - the calling code must be recompiled.

Rotvig replied:
True i will add the original method to the interface 👌☺️

sungam3r replied:
I personally do not really care 😄

sungam3r commented on Source/EasyNetQ/Consumer/ConsumerDispatcher.cs:
```diff
@@ -8,53 +8,78 @@ namespace EasyNetQ.Consumer
 
+        public bool IsDisposed { get; private set; }
         public void QueueAction(Action action)
         {
             Preconditions.CheckNotNull(action, "action");
-            queue.Add(action);
+            if (IsDisposed)
+                throw new ObjectDisposedException(nameof(ConsumerDispatcher));
+
+            transientActions.Enqueue(action);
+            autoResetEvent.Set();
+        }
```
` public void QueueAction(Action action) => QueueAction(action, false);`
